lib/elixir/lib/string.ex

```diff
@@ -1123,7 +1123,12 @@ defmodule String do
 
   """
   @spec replace(t, pattern | Regex.t, t, keyword) :: t
-  def replace(subject, pattern, replacement, options \\ []) when is_binary(replacement) do
+  def replace(subject, pattern, replacement, options \\ [])
+  def replace(subject, "", "", _), do: subject
+  def replace(subject, "", replacement, options) do
+    replace(subject, ~r//, replacement, options)
```

Comment:
I would rather implement this ourselves instead of dispatching to regex. Something like:

```elixir
defp intersperse(subject, replacement) do
  case next_grapheme(subject) do
    {current, rest} -> [current, replacement | intersperse(rest, replacement)]
    nil -> []
  end
end
```

And then you call `IO.iodata_to_binary [replacement | intersperse(subject, replacement)]` on this clause.

PS: I haven't tested this.

Reply:
I like your solution, `next_grapheme` is cool. 

The tests did not pass with your solution when the `global: false` option is provided, so I had to add another pattern match:

```elixir
assert String.replace("LICENSE","","x") == "xLxIxCxExNxSxEx"
assert String.replace("LICENSE","","x", global: true) == "xLxIxCxExNxSxEx"
assert String.replace("LICENSE","","x", global: false) == "xLICENSE"
# ^^^ did not pass
```

This is what I ended up with if we are not going to call out to `Regex.replace` via my original solution:

```elixir
  def replace(subject, pattern, replacement, options \\ [])
  def replace(subject, "", "", _), do: subject
  def replace(subject, "", replacement, global: false), do: replacement <> subject
  def replace(subject, "", replacement, _options) do
    IO.iodata_to_binary [replacement | intersperse(subject, replacement)]
  end
  def replace(subject, pattern, replacement, options) when is_binary(replacement) do
    if Regex.regex?(pattern) do
      Regex.replace(pattern, subject, replacement, global: options[:global])
    else
      opts = translate_replace_options(options)
      :binary.replace(subject, pattern, replacement, opts)
    end
  end

  defp intersperse(subject, replacement) do
    case next_grapheme(subject) do
      {current, rest} -> [current, replacement | intersperse(rest, replacement)]
      nil -> []
    end
  end
````

This passes all the tests, and although it is longer it is definitely more explicit. If that looks good to you I'll revise my commit. @josevalim

Reply:
Beautiful. We should not pattern match on keyword lists though, so it is probably better for you to do:

```elixir
  def replace(subject, "", replacement, options) do
    if Keyword.get(option, :global, true) do
      IO.iodata_to_binary [replacement | intersperse(subject, replacement)]
    else
      replacement <> subject
    end
  end
```

or something like that.

Reply:
@josevalim I've committed the changes you requested and added some documentation. Please let me know if you have any other concerns or anything else you would like me to change. Thanks!